dvc/dependency/repo.py

```diff
@@ -46,14 +47,37 @@ def _make_repo(self, **overrides):
         with external_repo(**merge(self.def_repo, overrides)) as repo:
             yield repo
 
-    def status(self):
-        with self._make_repo() as repo:
-            current = repo.find_out_by_relpath(self.def_path).info
+    # fileinfo is a dictionary containing "checksum" and "path"
+    def _get_checksum(self, updated=False):
+        rev_lock = None
+        if not updated:
+            rev_lock = self.def_repo.get(self.PARAM_REV_LOCK)
 
-        with self._make_repo(rev_lock=None) as repo:
-            updated = repo.find_out_by_relpath(self.def_path).info
+        try:
+            with self._make_repo(rev_lock=rev_lock) as repo:
+                path = os.path.join(repo.root_dir, self.def_path)
+
+                try:
+                    return repo.find_out_by_relpath(self.def_path).info["md5"]
+                except OutputNotFoundError:
+                    with repo.state:
+                        return self.repo.cache.local.get_checksum(
```

Comment:
No need to take `repo.state` if you are using `self.repo` anyways. Also, maybe you could fallthrough here, same as in your first `get` patch?

Reply:
Yes, and yes :)